### ISSUE TITLE
Allow static libraries to link dynamic frameworks

### ISSUE THREAD

**Issue by peter-tomaselli** [yonaskolb/XcodeGen#576]
We have an iOS app that we’re trying to migrate over to XcodeGen. Our `App` target links to many `Feature`s built as static libraries. Features can each link a common set of `CoreLibrary`s (for example), which are built as dynamic frameworks.

I’ve been digging in to the code and issue history here trying to understand why this [seems to be specifically disallowed](https://github.com/yonaskolb/XcodeGen/blob/master/Sources/XcodeGenKit/PBXProjGenerator.swift#L507) in XcodeGen. But I don’t understand why! It seems to be something related to the new build system perhaps? But as far as I can tell, whatever restriction used to be in place has been removed in Xcode 10.2, as this setup has been working fine for us.

For reference, if we comment out the line linked above, we get exactly the behavior we need. It also looks like there’s a `link` flag on `Dependency` that may be intended to override this behavior (eg. if the line above is changed to
```swift
if target.type == .staticLibrary && dependency.link != true { break }
```
we could also get what we need), but I’m not sure what the original design intention was there. 

Is either of these changes a viable approach? If so, I’d be happy to submit a PR. Thank you for your time!

**Comment by yonaskolb** [comment]
@brentleyjones do you have any comments as you worked on this? I'd be happy for either of these solutions, but as I'm unclear on what the restriction was for, maybe the second `link` flag solution is a safer default

**Comment by gomsu** [comment]
I have a similar problem where I have a framework that embeds multiple features that are static libraries. These static libraries are using a "core static library" for base features. I can't make the xcodegen to generate the "Link Binary With Libraries" build phase for the "feature static libraries" to link to the "core static library", so that the core would always get built first with any of the "feature static libraries". I can though add this build phase from the Xcode and add the "core static library"  to it and everything works after that, so it should be possible to generate it also.

**Comment by brentleyjones** [comment]
Static libraries can't link dynamic libraries (Xcode will correctly warn about it). You can depend on them, and as long as Xcode can find the headers and module maps you can use them in your static library, but they have to be linked to the same executable (which is possibly a dynamic library) that the static library ends up being linked to.

Our targets are almost exclusively static libraries, and we can depend on dynamic libraries just fine with XCG as it is now.

**Comment by vlozko** [comment]
> Static libraries can't link dynamic libraries or framework (Xcode will correctly warn about it).

This is simply incorrect. Otherwise you wouldn’t be able to build static libraries in a standalone fashion, period. Foundation is a dynamic framework, for example.

**Comment by brentleyjones** [comment]
@vlozko Can you provide me a sample Xcode project that does this?? When I linked static libraries to dynamic frameworks Xcode warned that it was incorrect to do so (and that it wouldn't actually link them). Static libraries can reference the symbols but depend on being linked to an executable that ultimately links to the dynamic framework. This makes sense, because static libraries are just an archive of object files.

**Comment by yonaskolb** [comment]
I believe this should be fixed by https://github.com/yonaskolb/XcodeGen/pull/635
Can someone confirm everything works ok now, and then close this

**Comment by yonaskolb** [comment]
Released in 2.7.0
